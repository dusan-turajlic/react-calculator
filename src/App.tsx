import React from 'react';
import Display from "./components/Display";
import Keyboard from "./components/Keyboard";
import { evaluateCalculation } from "./util";

interface AppState {
    input: string
}

const BASE_STATE_INPUT_VALUE = '0';

class App extends React.Component {
    state: AppState = {
        input: BASE_STATE_INPUT_VALUE
    };

    render() {
        return (
            <div className="app">
                <div className="p-absolute">
                    <Display input={ this.state.input }/>
                    <Keyboard onAnyKeyPress={ this.onInput.bind( this ) }/>
                </div>
            </div>
        );
    }

    onInput( event: React.MouseEvent, value: string ) {
        // Adds padding id non numerical characters are passed
        if ( [ '/', '*', '-', '+' ].includes( value ) ) {
            value = ' ' + value + ' ';
        }

        let currentInput = this.state.input;
        if ( this.state.input === BASE_STATE_INPUT_VALUE && value !== '.' ) {
            currentInput = '';
        }

        switch ( value ) {
            case '=':
                let calculatedValue = evaluateCalculation( this.state.input );
                this.setState( { input: calculatedValue } );
                break;
            case 'ce':
                this.setState( { input: BASE_STATE_INPUT_VALUE } );
                break;
            default:
                this.setState( { input: currentInput + value } );
                break;
        }

    }
}

export default App;
