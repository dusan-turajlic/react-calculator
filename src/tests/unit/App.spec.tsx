import React from 'react';
import { fireEvent, render, RenderResult } from '@testing-library/react';
import App from '../../App';

describe( 'App.spec', () => {
    let app: RenderResult;
    let display: HTMLInputElement | null;
    beforeEach( () => {
        app = render( <App/> );
        display = app.container.querySelector( 'input' );
    } );

    it( 'renders app', function () {
        expect( display ).toBeDefined();
        expect( app.container ).toMatchSnapshot();
    } );

    it( 'keyboard input displays data on display', async () => {
        const buttonOne = app.getByText( '1' );

        fireEvent.click( buttonOne );

        // @ts-ignore
        expect( display.value ).toBe( '1' );
    } );

    it( 'calculates value when pressing equals', async () => {
        const buttonNine = app.getByText( '9' );
        const buttonMinus = app.getByText( '-' );
        const buttonOne = app.getByText( '1' );
        const buttonEquals = app.getByText( '=' );

        fireEvent.click( buttonNine );
        fireEvent.click( buttonMinus );
        fireEvent.click( buttonOne );

        // @ts-ignore
        expect( display.value ).toBe( '9 - 1' );

        fireEvent.click( buttonEquals );

        // @ts-ignore
        expect( display.value ).toBe( '8' );
    } );
} );

