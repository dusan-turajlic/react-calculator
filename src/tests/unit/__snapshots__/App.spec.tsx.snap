// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.spec renders app 1`] = `
<div>
  <div
    class="app"
  >
    <div
      class="p-absolute"
    >
      <input
        class="form-input"
        readonly=""
        type="text"
        value="0"
      />
      <div
        class="container"
      >
        <div
          class="columns"
        >
          <button
            class="column col-3 "
          >
            1
          </button>
          <button
            class="column col-3 "
          >
            2
          </button>
          <button
            class="column col-3 "
          >
            3
          </button>
          <button
            class="column col-3 bg-gray"
          >
            /
          </button>
        </div>
        <div
          class="columns"
        >
          <button
            class="column col-3 "
          >
            4
          </button>
          <button
            class="column col-3 "
          >
            5
          </button>
          <button
            class="column col-3 "
          >
            6
          </button>
          <button
            class="column col-3 bg-gray"
          >
            *
          </button>
        </div>
        <div
          class="columns"
        >
          <button
            class="column col-3 "
          >
            7
          </button>
          <button
            class="column col-3 "
          >
            8
          </button>
          <button
            class="column col-3 "
          >
            9
          </button>
          <button
            class="column col-3 bg-gray"
          >
            -
          </button>
        </div>
        <div
          class="columns"
        >
          <button
            class="column col-3 "
          >
            .
          </button>
          <button
            class="column col-3 "
          >
            0
          </button>
          <button
            class="column col-3 "
          >
            ce
          </button>
          <button
            class="column col-3 bg-gray"
          >
            +
          </button>
        </div>
        <div
          class="columns"
        >
          <button
            class="column  bg-warning"
          >
            =
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
